service/src/main/kotlin/fi/espoo/evaka/messaging/message/MessageQueries.kt

```diff
@@ -0,0 +1,163 @@
+// SPDX-FileCopyrightText: 2017-2021 City of Espoo
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+package fi.espoo.evaka.messaging.message
+
+import fi.espoo.evaka.shared.Paged
+import fi.espoo.evaka.shared.db.Database
+import fi.espoo.evaka.shared.mapToPaged
+import fi.espoo.evaka.shared.withCountMapper
+import org.jdbi.v3.core.kotlin.mapTo
+import java.util.UUID
+
+fun Database.Read.getMessagesByAccount(
+    accountId: UUID,
+    pageSize: Int,
+    page: Int
+): Paged<MessageThread> {
+    val params = mapOf(
+        "accountId" to accountId,
+        "offset" to (page - 1) * pageSize,
+        "pageSize" to pageSize
+    )
+
+    // language=SQL
+    val sql = """
+WITH thread_participation AS (
+    SELECT mt.id AS thread_id,
+           mp.message_id,
+           mp.account_id,
+           sender,
+           read_at
+    FROM message_participants mp
+             JOIN message m ON mp.message_id = m.id
+             JOIN message_thread_messages mtm ON mtm.message_id = m.id
+             JOIN message_thread mt ON mtm.thread_id = mt.id
+    WHERE EXISTS(SELECT 1
+                 FROM message_participants participants
+                 WHERE participants.message_id = m.id
+                   AND participants.account_id = :accountId)
+),
+     threads(id, type, count) AS (
+         SELECT thread.id,
+                thread.message_type AS type,
+                count(*) OVER ()    AS count
+         FROM message_thread thread
+         WHERE EXISTS(SELECT 1
+                      FROM thread_participation
+                      WHERE thread_participation.thread_id = thread.id)
+         ORDER BY thread.created DESC
+         LIMIT :pageSize OFFSET :offset
+     ),
+     messages AS (
+         SELECT sender.thread_id,
+                m.id,
+                m.content,
+                m.title,
+                m.created,
+                sender.account_id AS sender_id,
+                name_view.account_name as sender_name
+         FROM message m
+         JOIN thread_participation sender ON sender.message_id = m.id AND sender.sender = TRUE
+         JOIN message_account_name_view name_view ON sender.account_id = name_view.id
+         ORDER BY m.created ASC
+     ),
```

Comment:
I don't personally like using a large number of sub-query aliases like this. However, it is certainly in line with how things are done in this project.

I would love it if we could reduce SQL complexity in favour of Kotlin complexity, but that would probably result in a performance hit. 🤷

Reply:
Yeah, this approach gets messy fast.
What we could do here is break this up two separate queries: fetch threads with paging first and their messages in another query. What do you think?